### PR TITLE
allow cancelling the zoom function with ESC

### LINKED ISSUE #49804: Allow to cancel the zoom function
### Feature description

Once the user has startet to zoom in, currently there is no way to cancel this (e.g. by pressing the ESCAPE key). It would be great to be able to do so.

In the following state when pressing ESC nothing happens such that the user is forced to zoom in:
Bildschirmfoto vom 2022-08-16 13-47-20

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by edaigle**
As described in #49804, there is currently no way to cancel a zoom operation once
the user has begun dragging. This commit makes the ESC key cancel an in-progress
zoom operation, leaving the canvas unchanged.

Closes #49804

### CODE REVIEW

File: src/gui/qgsmaptoolzoom.h [hunk 1]
```diff
@@ -58,6 +58,9 @@ class GUI_EXPORT QgsMapToolZoom : public QgsMapTool
     //! Flag to indicate a map canvas drag operation is taking place
     bool mDragging;
 
+    //! Flag to indicate the user has canceled the current zoom operation
+    bool canceled;
```
Comment (nyalldawson) [inline src/gui/qgsmaptoolzoom.h:62]:
```suggestion
    bool mCanceled = false;
```

File: src/gui/qgsmaptoolzoom.cpp [hunk 2]
```diff
@@ -85,6 +85,7 @@ void QgsMapToolZoom::canvasPressEvent( QgsMapMouseEvent *e )
     return;
 
   mZoomRect.setRect( 0, 0, 0, 0 );
+  mZoomRect.setTopLeft( e->pos() );
```
Comment (uclaros) [inline src/gui/qgsmaptoolzoom.cpp:88]: Why is this needed?
This will set a rect from 0,0 to the mouse pos.
If anything, we could also set bottom right to mouse pos so that we're starting with an empty rect at mouse click pos.

File: src/gui/qgsmaptoolzoom.h (hunk 1 above)
Comment (edaigle) [reply src/gui/qgsmaptoolzoom.h]: Makes sense, change made. Thanks!

File: src/gui/qgsmaptoolzoom.cpp (hunk 2 above)
Comment (edaigle) [reply src/gui/qgsmaptoolzoom.cpp:87]: Ah, that was my intention but missed changing the line - thanks for catching it! We now start with an empty rect as mouse click pos.

File: src/gui/qgsmaptoolzoom.cpp [hunk 3]
```diff
@@ -172,4 +180,11 @@ void QgsMapToolZoom::keyReleaseEvent( QKeyEvent *e )
   {
     setZoomMode( mNativeZoomOut );
   }
+
+  if ( e->key() == Qt::Key_Escape && mDragging )
+  {
+    canceled = true;
+    delete mRubberBand;
+    mRubberBand = nullptr;
+  }
```
[the remark this answers is not in the document]
Comment (edaigle) [reply src/gui/qgsmaptoolzoom.cpp]: For some reason `keyPressEvent()` doesn't fire while the mouse is held down (as it is when the user is dragging to zoom, and then needs to cancel it): 

` // key press events are not caught wile the mouse is pressed
  // this is detected in map canvas move
`

I've tested that this is still the case - unless there's a workaround I've missed (quite possible) I believe it has to be on `keyReleaseEvent()` to work around it.

File: src/gui/qgsmaptoolzoom.h [hunk 4]
```diff
@@ -58,6 +58,9 @@ class GUI_EXPORT QgsMapToolZoom : public QgsMapTool
     //! Flag to indicate a map canvas drag operation is taking place
     bool mDragging;
 
+    //! Flag to indicate the user has canceled the current zoom operation
+    bool mCanceled;
```
Comment (lbartoletti) [inline src/gui/qgsmaptoolzoom.h:62]:
```suggestion
    bool mCanceled = false;
```

As mentioned by Nyall, you must initialize it at false.

File: src/gui/qgsmaptoolzoom.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (uclaros) [reply src/gui/qgsmaptoolzoom.cpp]: OK then, apparently this is also how it works for the selection tool too.

**Comment by edaigle** [comment]
As far as I know this is ready to be merged:)

**Review by nyalldawson — APPROVED** [review]
